chopper/lib/src/authenticator.dart

```diff
@@ -2,12 +2,32 @@ import 'dart:async';
 
 import 'package:chopper/chopper.dart';
 
-/// This method should return a [Request] that includes credentials to satisfy an authentication challenge received in
+/// This method should return a [Request] that includes credentials to satisfy
+/// an authentication challenge received in
 /// [response]. It should return `null` if the challenge cannot be satisfied.
+///
+/// Optionally, you can override either [onAuthenticationSuccessful] or
+/// [onAuthenticationFailed] in order to listen to when a particular
+/// authentication request succeeds or fails. You can also use it in order
+/// to reset or mutate your instance's internal state for the purposes
+/// of keeping track of the number of retries made to authenticate a
+/// request.
 abstract class Authenticator {
   FutureOr<Request?> authenticate(
     Request request,
     Response response, [
     Request? originalRequest,
   ]);
+
+  FutureOr<void> onAuthenticationSuccessful(
+    Request request,
+    Response response, [
+    Request? originalRequest,
+  ]) {}
```

Comment:
I'm not sure I like empty methods. Can you maybe change it to a getter, i.e.

```dart
import 'dart:async';

import 'package:chopper/chopper.dart';

typedef AuthenticationCallback = FutureOr<void> Function(
  Request request,
  Response response, [
  Request? originalRequest,
]);

/// This method should return a [Request] that includes credentials to satisfy
/// an authentication challenge received in
/// [response]. It should return `null` if the challenge cannot be satisfied.
///
/// Optionally, you can override either [onAuthenticationSuccessful] or
/// [onAuthenticationFailed] in order to listen to when a particular
/// authentication request succeeds or fails. You can also use it in order
/// to reset or mutate your instance's internal state for the purposes
/// of keeping track of the number of retries made to authenticate a
/// request.
abstract class Authenticator {
  FutureOr<Request?> authenticate(
    Request request,
    Response response, [
    Request? originalRequest,
  ]);

  AuthenticationCallback? get onAuthenticationSuccessful => null;

  AuthenticationCallback? get onAuthenticationFailed => null;
}
```

Reply:
On it.

Reply:
Done.

Reply:
Can you maybe fix the docs? Whatever you add here will be displayed [in the API section on pub.dev](https://pub.dev/documentation/chopper/latest/chopper/Authenticator-class.html).

Maybe something like this

```dart
import 'dart:async';

import 'package:chopper/chopper.dart';

/// A callback that is called when an authentication challenge is received.
typedef AuthenticationCallback = FutureOr<void> Function(
  Request request,
  Response response, [
  Request? originalRequest,
]);

/// An [Authenticator] is used to handle authentication challenges.
abstract class Authenticator {
  /// This method should return a [Request] that includes credentials to satisfy
  /// an authentication challenge received in [response].
  /// It should return `null` if the challenge cannot be satisfied.
  FutureOr<Request?> authenticate(
    Request request,
    Response response, [
    Request? originalRequest,
  ]);

  /// Override to listen to an authentication success.
  AuthenticationCallback? get onAuthenticationSuccessful => null;

  /// Override to listen to an authentication failure.
  AuthenticationCallback? get onAuthenticationFailed => null;
}

```

Reply:
Done.